storage: panic: error resolving intent at /Table/55/3/4{400/"G"/0-999/"H"/0/NULL} on end transaction [COMMITTED]: cannot read undeclared span /{Table/55/3/4999/"H"/0/NULL-Min}

Seen [in CI](https://teamcity.cockroachdb.com/viewLog.html?buildId=449849&buildTypeId=Cockroach_UnitTests_Testrace&tab=buildLog).

We are trying to resolve an intent in `EndTransaction` which lives at

`/Table/55/3/4400/"G"/0-/Table/55/3/4999/"H"/0/NULL`

but we only declared a key range that has that key as its endpoint (i.e. it's not contained):

`/Table/55/3/4999/"H"/0/NULL`

The offending EndTransaction seems to be from a schema change truncation batch. As a result, I looked into https://github.com/cockroachdb/cockroach/blob/e185c344507ad0a8ca8bd6d8adbdb2c017f63afd/pkg/storage/replica_command.go#L527-L535

but wasn't able to find anything. Note that we declare the whole intent span, not its truncation to the range, so the endpoint appears to actually have been touched unintentionally. Maybe there's a bug in `MVCCResolveWriteIntentRangeUsingIter`... oh, I see something I think. That code is essentially

```go
nextKey := key
for {
  iter.Seek(key)
  if !iter.OK() || !key.InRange() { break }
  key := iter.UnsafeKey()
  nextKey = key.Next()
}
```

which means that we'll actually cann `Seek` for a key that's already outside of the range, for example when `range = [a, b\x00)` and previous seek went to `b` (which would be valid): then we would set `nextKey = b\x00` and call seek again, which is now illegal since we promised to stay within `[a, b\x00)`.
I'll whip up a unit test and fix.

```
goroutine 1335 [select]:
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas(0xc4212fa780, 0x4002280, 0xc4215c2c80, 0xc4212fa7d0, 0x2a, 0xc42355b630, 0x3, 0x3, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1256 +0x5f2
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC(0xc4212fa780, 0x4002280, 0xc4215c2c80, 0x2a, 0xc42355b630, 0x3, 0x3, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:383 +0x2f0
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendSingleRange(0xc4212fa780, 0x4002280, 0xc4215c2c80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420a020e0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:447 +0x1ab
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch(0xc4212fa780, 0x4002280, 0xc4215c2c80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420a020e0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1025 +0x403
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges(0xc4212fa780, 0x4002280, 0xc4215c2c80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420a020e0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:701 +0x1d1e
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send(0xc4212fa780, 0x4002280, 0xc4215c2c80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420f33dc0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:622 +0x53a
github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send(0xc420fe8fc0, 0x4002280, 0xc4215c2c80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420f33dc0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:444 +0x259
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).send(0xc4208ca650, 0x4002280, 0xc4215c2c80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420f33dc0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:546 +0x31f
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Send(0xc4214a1500, 0x4002280, 0xc4215c2c80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:951 +0x589
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).sendEndTxnReq(0xc4214a1500, 0x4002280, 0xc4215c2c80, 0xc42403a501, 0x0, 0xc4214a1538)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:671 +0x293
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).commit(0xc4214a1500, 0x4002280, 0xc4215c2c80, 0xc4214a1538, 0x2d378c0)
[17:27:37] :		 [TestImportStmt/schema-in-file-sstsize-dist] [Test Output]
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:518 +0x74
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Commit(0xc4214a1500, 0x4002280, 0xc4215c2c80, 0xc423498f98, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:550 +0x4d
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc4214a1500, 0x4002280, 0xc4215c2c80, 0xc4215c0101, 0xc421490570, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:764 +0x65f
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn(0xc4208ca650, 0x4002280, 0xc4215c2c80, 0xc421c07560, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:507 +0x17b
github.com/cockroachdb/cockroach/pkg/sql.truncateTableInChunks(0x4002340, 0xc420dbdef0, 0xc423154900, 0xc4208ca650, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/truncate.go:349 +0x331
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).maybeAddDropRename(0xc4208ca640, 0x4002340, 0xc420dbdef0, 0xc421643f00, 0xc423408b20, 0xc423154900, 0xc423408b20, 0xc41ffe75e6, 0xc41ffe75e5)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:309 +0x1c9
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).exec(0xc4208ca640, 0x4002340, 0xc420dbdef0, 0xc420431700, 0x0, 0x0, 0x0, 0x0, 0x8345668, 0xed1cb4065, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:415 +0x286
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChangeManager).Start.func1(0x4002340, 0xc42161d110)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:1116 +0x8cf
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc420bda760, 0xc4214b2fc0, 0xc42039b080)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x161
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xbb
```

and here's the stack trace:

```
[17:27:37]F:	 [Step 3/4] TestImportStmt/schema-in-file-sstsize-dist
[17:27:37] :		 [TestImportStmt/schema-in-file-sstsize-dist] [Test Output]
I171219 17:16:26.167805 2142 sql/event_log.go:113  [client=127.0.0.1:45880,user=root,n1] Event: "create_database", target: 60, info: {DatabaseName:csv5 Statement:CREATE DATABASE csv5 User:root}
I171219 17:16:26.220538 883 storage/replica_command.go:1231  [split,n1,s1,r55/1:/{Table/59/3/4…-Max}] initiating a split of this range at key /Table/60 [r59]
I171219 17:16:26.541336 834 storage/replica_proposal.go:195  [n1,s1,r55/1:/{Table/59/3/4…-Max}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1513703766.081625393,0 following repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1513703766.081625393,0
I171219 17:16:26.639244 1224 storage/replica_proposal.go:195  [n2,s2,r39/3:/Table/5{5/3/4999…-6}] new range lease repl=(n2,s2):3 start=1513703786.630564551,0 epo=1 pro=1513703786.630590731,0 following repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1513703766.081625393,0
I171219 17:16:26.860500 1161 gossip/gossip.go:1293  [n2] node has connected to cluster via gossip
I171219 17:16:26.861504 1161 storage/stores.go:351  [n2] wrote 2 node addresses to persistent storage
I171219 17:16:27.053016 1252 server/status/runtime.go:219  [n2] runtime stats: 845 MiB RSS, 537 goroutines, 45 MiB/18 MiB/80 MiB GO alloc/idle/total, 125 MiB/155 MiB CGO alloc/total, 19303.56cgo/sec, 2.31/0.11 %(u/s)time, 0.02 %gc (19x)
I171219 17:16:27.593017 1360 server/status/runtime.go:219  [n3] runtime stats: 876 MiB RSS, 661 goroutines, 55 MiB/5.9 MiB/80 MiB GO alloc/idle/total, 152 MiB/179 MiB CGO alloc/total, 18684.56cgo/sec, 2.32/0.11 %(u/s)time, 0.02 %gc (19x)
I171219 17:16:28.249379 1352 gossip/gossip.go:1293  [n3] node has connected to cluster via gossip
I171219 17:16:28.250298 1352 storage/stores.go:351  [n3] wrote 2 node addresses to persistent storage
panic: error resolving intent at /Table/55/3/4{400/"G"/0-999/"H"/0/NULL} on end transaction [COMMITTED]: cannot read undeclared span /{Table/55/3/4999/"H"/0/NULL-Min}
declared:
0 1: /Local/Range/Table/55/3{/RangeDescriptor-}
1 0: /Table/55/3/4{400/"G"/0-999/"H"/0/NULL}
1 1: /Local/Range/{Table/55/3/4400/"G"/0/Transaction/"ff22b557-33ff-4dc2-84d6-26b38dc9fe62"-"\x12\xbf\x8b\xf7\x110\x12G\x00\xff\x01\x88\x00\x01txn-\xff\"\xb5W3\xffM\u0084\xd6&\xb3\x8d\xc9\xfeb\x00"/err:uuid: UUID must be exactly 16 bytes long, got 17 bytes}
1 1: /Local/RangeID/42/r/AbortSpan/"ff22b557-33ff-4dc2-84d6-26b38dc9fe62"{-}
 [recovered]
	panic: error resolving intent at /Table/55/3/4{400/"G"/0-999/"H"/0/NULL} on end transaction [COMMITTED]: cannot read undeclared span /{Table/55/3/4999/"H"/0/NULL-Min}
declared:
0 1: /Local/Range/Table/55/3{/RangeDescriptor-}
1 0: /Table/55/3/4{400/"G"/0-999/"H"/0/NULL}
1 1: /Local/Range/{Table/55/3/4400/"G"/0/Transaction/"ff22b557-33ff-4dc2-84d6-26b38dc9fe62"-"\x12\xbf\x8b\xf7\x110\x12G\x00\xff\x01\x88\x00\x01txn-\xff\"\xb5W3\xffM\u0084\xd6&\xb3\x8d\xc9\xfeb\x00"/err:uuid: UUID must be exactly 16 bytes long, got 17 bytes}
1 1: /Local/RangeID/42/r/AbortSpan/"ff22b557-33ff-4dc2-84d6-26b38dc9fe62"{-}
 [recovered]
	panic: error resolving intent at /Table/55/3/4{400/"G"/0-999/"H"/0/NULL} on end transaction [COMMITTED]: cannot read undeclared span /{Table/55/3/4999/"H"/0/NULL-Min}
declared:
0 1: /Local/Range/Table/55/3{/RangeDescriptor-}
1 0: /Table/55/3/4{400/"G"/0-999/"H"/0/NULL}
1 1: /Local/Range/{Table/55/3/4400/"G"/0/Transaction/"ff22b557-33ff-4dc2-84d6-26b38dc9fe62"-"\x12\xbf\x8b\xf7\x110\x12G\x00\xff\x01\x88\x00\x01txn-\xff\"\xb5W3\xffM\u0084\xd6&\xb3\x8d\xc9\xfeb\x00"/err:uuid: UUID must be exactly 16 bytes long, got 17 bytes}
1 1: /Local/RangeID/42/r/AbortSpan/"ff22b557-33ff-4dc2-84d6-26b38dc9fe62"{-}


goroutine 7109 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc420cec510, 0x4002340, 0xc421b37260)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:181 +0x131
panic(0x2953660, 0xc420282fc0)
	/usr/local/go/src/runtime/panic.go:491 +0x2a2
github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send.func1(0xc4242c2dd0, 0xc4242c2e48, 0x1501c19390333d38, 0x0, 0xc4242c2e40, 0xc420efe400)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2464 +0x763
panic(0x2953660, 0xc420282fc0)
	/usr/local/go/src/runtime/panic.go:491 +0x2a2
github.com/cockroachdb/cockroach/pkg/storage.resolveLocalIntents(0x4002340, 0xc421890fc0, 0xc420ffd3e0, 0x7f91186db000, 0xc42355bb30, 0xc420da9730, 0xc422f84670, 0xa, 0x10, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:546 +0x59a
github.com/cockroachdb/cockroach/pkg/storage.evalEndTransaction(0x4002340, 0xc421890fc0, 0x7f91186db000, 0xc42355bb30, 0x4015500, 0xc420e2ee70, 0x1501c19387c1f5db, 0x0, 0x100000001, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:400 +0x1736
github.com/cockroachdb/cockroach/pkg/storage.evaluateCommand(0x4002340, 0xc421890fc0, 0xc4210dcab8, 0x8, 0x0, 0x7f91186db000, 0xc42355bb30, 0x4015500, 0xc420e2ee70, 0xc420da9730, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:114 +0x343
github.com/cockroachdb/cockroach/pkg/storage.evaluateBatch(0x4002340, 0xc421890fc0, 0xc4210dcab8, 0x8, 0x7f91186db000, 0xc42355bb30, 0x4015500, 0xc420e2ee70, 0xc420da9730, 0x1501c19387c1f5db, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5319 +0x431
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).evaluateTxnWriteBatch(0xc42022f880, 0x4002340, 0xc421890fc0, 0xc4210dcab8, 0x8, 0x1501c19387c1f5db, 0x0, 0x100000001, 0x1, 0x2a, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5121 +0x328
[17:27:37] :		 [TestImportStmt/schema-in-file-sstsize-dist] [Test Output]
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).evaluateProposalInner(0xc42022f880, 0x4002340, 0xc421890fc0, 0xc4210dcab8, 0x8, 0x1501c19387c1f5db, 0x0, 0x100000001, 0x1, 0x2a, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4957 +0x186
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).evaluateProposal(0xc42022f880, 0x4002340, 0xc421890fc0, 0xc4210dcab8, 0x8, 0x1501c19387c1f5db, 0x0, 0x100000001, 0x1, 0x2a, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2866 +0x136
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).requestToProposal(0xc42022f880, 0x4002340, 0xc421890fc0, 0xc4210dcab8, 0x8, 0x1501c19387c1f5db, 0x0, 0x100000001, 0x1, 0x2a, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2823 +0x2eb
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).propose(0xc42022f880, 0x4002340, 0xc421890fc0, 0x0, 0x0, 0xc4234053f0, 0x100000001, 0x1, 0x0, 0xc423405400, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2995 +0x5a9
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).tryExecuteWriteBatch(0xc42022f880, 0x4002340, 0xc421890fc0, 0x1501c19387c1f5db, 0x0, 0x100000001, 0x1, 0x2a, 0x0, 0xc421431180, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2724 +0x3d7
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).executeWriteBatch(0xc42022f880, 0x4002340, 0xc421890fc0, 0x1501c19387c1f5db, 0x0, 0x100000001, 0x1, 0x2a, 0x0, 0xc421431180, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2578 +0xbf
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).Send(0xc42022f880, 0x4002340, 0xc421890fc0, 0x1501c19387c1f5db, 0x0, 0x100000001, 0x1, 0x2a, 0x0, 0xc421431180, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1901 +0x7aa
github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send(0xc420efe400, 0x4002340, 0xc421b372f0, 0x1501c19387c1f5db, 0x0, 0x100000001, 0x1, 0x2a, 0x0, 0xc421431180, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2571 +0x7a6
```

--- a/pkg/storage/batch_spanset_test.go
+++ b/pkg/storage/batch_spanset_test.go
@@ -17,13 +17,18 @@ package storage
 
 import (
 	"bytes"
+	"math"
 	"reflect"
 	"testing"
 
+	"golang.org/x/net/context"
+
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage/engine"
+	"github.com/cockroachdb/cockroach/pkg/storage/engine/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/storage/spanset"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/pkg/errors"
 )
@@ -185,5 +190,51 @@ func TestSpanSetBatch(t *testing.T) {
 	iter.SeekReverse(insideKey)
 	if ok, err := iter.Valid(); !ok {
 		t.Fatalf("expected valid iterator, err=%v", err)
+	}
+}
+
+// TestSpanSetMVCCResolveWriteIntentRangeUsingIter verifies that
+// MVCCResolveWriteIntentRangeUsingIter does not stray outside of the passed-in
+// key range (which it only used to do in this corner case tested here).
+//
+// See #20894.
+func TestSpanSetMVCCResolveWriteIntentRangeUsingIter(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	eng := engine.NewInMem(roachpb.Attributes{}, 10<<20)
+	defer eng.Close()
+
+	ctx := context.Background()
+
+	value := roachpb.MakeValueFromString("irrelevant")
+
+	if err := engine.MVCCPut(
+		ctx,
+		eng,
+		nil, // ms
+		roachpb.Key("b"),
+		hlc.Timestamp{WallTime: 10}, // irrelevant
+		value,
+		nil, // txn
+	); err != nil {
+		t.Fatal(err)
+	}
+
+	var ss spanset.SpanSet
+	ss.Add(spanset.SpanReadWrite, roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("b\x00")})
+
+	batch := spanset.NewBatch(eng.NewBatch(), &ss)
+	defer batch.Close()
+
+	iterAndBuf := engine.GetIterAndBuf(batch)
+	defer iterAndBuf.Cleanup()
+
+	intent := roachpb.Intent{
+		Span:   roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("b\x00")},
+		Txn:    enginepb.TxnMeta{}, // unused
+		Status: roachpb.PENDING,
+	}
+
+	if _, err := engine.MVCCResolveWriteIntentRangeUsingIter(ctx, batch, iterAndBuf, nil /* ms */, intent, math.MaxInt64); err != nil {
+		t.Fatal(err)
 	}
 }

--- a/pkg/storage/engine/mvcc.go
+++ b/pkg/storage/engine/mvcc.go
@@ -2152,8 +2152,13 @@ func MVCCResolveWriteIntentRangeUsingIter(
 			num++
 		}
 
-		// nextKey is already a metadata key.
+		// nextKey is already a metadata key...
 		nextKey.Key = key.Key.Next()
+		if !nextKey.Less(encEndKey) {
+			// ... but we don't want to Seek to a key outside of the range as we validate
+			// those span accesses (see TestSpanSetMVCCResolveWriteIntentRangeUsingIter).
+			break
+		}
 	}
 
 	return num, nil

